File: AlgorithmStudy/9주차/BOJ10026/김재환10026.java
import java.io.*;
import java.util.*;

public class Main {

    static int N;
    static char[][] map;
    static boolean[][] visited;
    
    // 상, 하, 좌, 우
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};

    static void bfs(int x, int y, boolean colorBlind) {
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{x, y});
        visited[x][y] = true;
        char startColor = map[x][y];

        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int cx = cur[0];
            int cy = cur[1];

            for (int i = 0; i < 4; i++) {
                int nx = cx + dx[i];
                int ny = cy + dy[i];

                // 범위 체크
                if (nx < 0 || nx >= N || ny < 0 || ny >= N) continue;
                if (visited[nx][ny]) continue;

                if (colorBlind) {
                    // 적록색약: R/G 동일
                    if ((startColor == 'R' || startColor == 'G') &&
                        (map[nx][ny] == 'R' || map[nx][ny] == 'G')) {
                        visited[nx][ny] = true;
                        q.add(new int[]{nx, ny});
                    }
                    // B는 B끼리만
                    else if (startColor == 'B' && map[nx][ny] == 'B') {
                        visited[nx][ny] = true;
                        q.add(new int[]{nx, ny});
                    }
                } else {
                    // 정상 시야
                    if (map[nx][ny] == startColor) {
                        visited[nx][ny] = true;
                        q.add(new int[]{nx, ny});
                    }
                }
            }
        }
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        map = new char[N][N];

        for (int i = 0; i < N; i++) {
            map[i] = br.readLine().toCharArray();
        }

        // 정상 시야
        visited = new boolean[N][N];
        int normal = 0;

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (!visited[i][j]) {
                    bfs(i, j, false);
                    normal++;
                }
            }
        }

        // 적록색약
        visited = new boolean[N][N];
        int colorBlind = 0;

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (!visited[i][j]) {
                    bfs(i, j, true);
                    colorBlind++;
                }
            }
        }

        System.out.println(normal + " " + colorBlind);
    }
}

File: AlgorithmStudy/9주차/BOJ10026/김재환10026.py
from collections import deque
import sys
input = sys.stdin.readline

n = int(input())
graph = [list(input().strip()) for _ in range(n)]

# 상, 하, 좌, 우 이동
dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

def bfs(x, y, color_blind):
    """
    x, y : 시작 좌표
    color_blind : True면 적록색약, False면 정상
    """
    queue = deque()
    queue.append((x, y))
    visited[x][y] = True
    start_color = graph[x][y]

    while queue:
        cx, cy = queue.popleft()

        for i in range(4):
            nx = cx + dx[i]
            ny = cy + dy[i]

            # 범위 체크 + 방문 여부
            if 0 <= nx < n and 0 <= ny < n and not visited[nx][ny]:

                if color_blind:
                    # 적록색약인 경우
                    # R과 G는 같은 색으로 본다
                    if (start_color in "RG" and graph[nx][ny] in "RG") or \
                       (start_color == "B" and graph[nx][ny] == "B"):
                        visited[nx][ny] = True
                        queue.append((nx, ny))
                else:
                    # 정상 시야: 색이 정확히 같아야 함
                    if graph[nx][ny] == start_color:
                        visited[nx][ny] = True
                        queue.append((nx, ny))

# -------- 정상 시야 --------
visited = [[False] * n for _ in range(n)]
normal_count = 0

for i in range(n):
    for j in range(n):
        if not visited[i][j]:
            bfs(i, j, False)
            normal_count += 1

# -------- 적록색약 --------
visited = [[False] * n for _ in range(n)]
color_blind_count = 0

for i in range(n):
    for j in range(n):
        if not visited[i][j]:
            bfs(i, j, True)
            color_blind_count += 1

print(normal_count, color_blind_count)
